### DISCUSSION TITLE
[RESOLVIDO] Notificação de Update

### DISCUSSION THREAD

**Discussion by John-Rocha** [Flutterando/forum#12] [Perguntas e Respostas] [answered]
Alguém sabe me dizer se há um package ou como faz para que o app verifique se tem versão nova na loja e "force" o usuário a atualizá-lo?

**Comment by GabrielCR99** [comment]
Existem alguns packages bons pra poder verificar se há uma nova atualização no app, e então "forçar" a atualização. 
Atualmente, a comunidade passou a cuidar do package new_version, visto que esse estava desatualizado. A nova versão e a mais estável e recomendade, é essa https://pub.dev/packages/new_version_plus

**Comment by paulobreim** [comment]
Execute os comandos:
flutter pub add update_available
flutter pub add url_launcher

e ai executa o .dart abaixo com  verificaNovaVersao(context);

import 'package:update_available/update_available.dart';
import 'package:url_launcher/url_launcher.dart';

import 'package:flutter/material.dart';
import 'package:update_available/update_available.dart';
import 'package:url_launcher/url_launcher.dart';

Future<bool> novaVersao(var context) async {
  var quer = false;
  Widget cancelaButton = TextButton(
    child: const Text('Quero sim', textScaleFactor: 1.0),
    onPressed: () {
      quer = true;
      Navigator.of(context).pop();
    },
  );
  Widget continuaButton = TextButton(
    child: const Text('Depois', textScaleFactor: 1.0),
    onPressed: () {
      quer = false;
      Navigator.of(context).pop();
    },
  );

  await showDialog(
    context: context,
    barrierDismissible: true,
    builder: (context) {
      return AlertDialog(
        title: const Text('Nova Versão', textScaleFactor: 1.0),
        content: const Text(
            'Tem uma nova versão do **xxxx**, quer atualizar?',
            textScaleFactor: 1.0),
        actions: <Widget>[
          cancelaButton,
          continuaButton,
        ],
        backgroundColor: Colors.yellow,
        elevation: 24,
        shape: const RoundedRectangleBorder(
            borderRadius: BorderRadius.all(Radius.circular(20.0))),
      );
    },
  );
  return quer;
}

Future<void> _launchInBrowser(String url) async {
  if (await canLaunchUrl(Uri.parse(url))) {
    await launchUrl(
      Uri.parse(url),
      // forceSafariVC: false,
      // forceWebView: false,
      // headers: <String, String>{'my_header_key': 'my_header_value'},
    );
  } else {
    throw 'Could not launch $url';
  }
}

bool perguntou = false;

Future<bool> verificaNovaVersao(var context) async {
  if (perguntou) {
    return true;
  }
  perguntou = true;
  Availability av = await getUpdateAvailability();
  switch (av) {
    case UnknownAvailability:
      //   print('putz, não sei se tem versão nova.');
      break;
    case NoUpdateAvailable:
      //  print('não tem versão nova.');
      break;
    case UpdateAvailable:
      //  print('Tem versão nova.');
      if (await novaVersao(context)) {
        await _launchInBrowser(
            'https://play.google.com/store/apps/details?id=**com.xxx.yyy**);
      }
      break;
    default:
      // print('resposta inesperada. $av');
      break;
  }

  return true;
}

**Reply by mateusfccp** [reply]
@paulobreim Poderia formatar seu código? Vai ajudar a entender melhor.

Você consegue fazer isso colocando todo seu código entre backticks, conforme exemplo:

<img width="132" alt="image" src="https://user-images.githubusercontent.com/4605213/190149532-e6586c04-60bf-4640-9262-11c7a3a36a3b.png">

Resultado:

```dart
void main() {
  // Do something
}
```

**Reply by paulobreim** [reply]
[novaversao.dart.txt](https://github.com/Flutterando/forum/files/9567668/novaversao.dart.txt)
coloquei em txt.

Chamando a função m verificaNovaVersao(context); se tiver nova versão ele abre uma janela perguntando se o usuário quer atualizar.

**Comment by jacobaraujo7** [comment]
Resposta rápida, NÃO TEM!
Todos os packages que dizem fazer isso na realidade estão apenas fazendo Crawler do site da Google Play e App Store, então muito cuidado ao adicionar eles.

O que recomendo é usar o Firebase Remote Config e settar a ultima versão do seu app lá, fazendo com que apareça uma tela em vez do widget inicial obrigando o usuário a atualizar o app. Esse link pode ser disparado usando o url_launcher por exemplo.

Espero ter ajudade :)

**Reply by GabrielCR99** [reply]
Uma dúvida que tenho em relação ao uso do Firebase, se o escopo não permite (por quaisquer motivos que seja) o uso do do Remote Config, como seria feita a verificação do update?

**Reply by paulobreim** [reply]
A sugestão que dei acima funciona e eu uso em todos meus apps.
A Google sabe disso e or isso não deve mexer porque vai quebrar.

**Reply by jacobaraujo7** [reply]
> Uma dúvida que tenho em relação ao uso do Firebase, se o escopo não permite (por quaisquer motivos que seja) o uso do do Remote Config, como seria feita a verificação do update?

Acredito que um consumo direto no seu backend ao iniciar a aplicação é bom tb

**Comment by renanzdm** [comment, marked as answer]
O Android possui uma Api nativa que pode fazer isso para voce, porem nao sei se ja tem algum package que faz dessa forma no flutter pode dar uma olhada aqui https://developer.android.com/guide/playcore/in-app-updates?hl=pt-br. Porem para Ios nao temos nada nativo

**Reply by jacobaraujo7** [reply]
Acho que essa é a resposta correta @John-Rocha

**Reply by jacobaraujo7** [reply]
Só adicionalmente, olhando o código feito do [package](https://pub.dev/packages/update_available) informado pelo @paulobreim usa essa abordagem no android. Porém no iOS é Crawler.

**Reply by renanzdm** [reply]
Sim, nao havia visto esse package ainda. Mas ele usa isso no Android mesmo, que ao meu ver e muito mais elegante. Acho uma pena no ios nao termos essa possibilidade, mas acredito que seja por que geralemte no ios as atualizacoes automaticas ja sao o padrao

**Reply by mateusfccp** [reply]
Infelizmente a bostapple não disponibiliza API para verificar updates no iOS. Meu package usa uma API da própria Apple, mas é uma API REST cujo propósito principal não é esse, então não sabemos se um dia ela pode simplesmente mudar ou parar de funcionar.

Dito isso, o pacote tem funcionado desde 2020.

**Comment by dtodt** [comment]
Tem um Package que tenho utilizado, que facilita bastante o controle de atualização [upgrader](https://pub.dev/packages/upgrader).

Apesar de contar com os crawlers, ele possui uma opção de usar XML como fonte verdade.
O que nos possibilita um controle fino de quando solicitar uma atualização ou quando obrigar uma atualização.

O package suporta Android e iOS e a abordagem dele é semelhante à usar o url_launcher em conjunto com o update_available, só que quando em conjunto com o XML, crawlers não são necessários.

Temos usado ele desta forma, e o XML colocamos em um repositório git publico, mas também daria pra colocar ele em um remote config, e fazer o controle de forma mais granular.

**Comment by John-Rocha** [comment]
Senhores, obrigado pela discussão e dicas. Foram bem avaliadas.
Nós conseguimos resolver essa demando usando o Firebase Config. Foi a maneira mais prática e segura que avaliamos.

**Comment by paulobreim** [comment]
Funciona mesmo. Us em vários aplicativos desde a primeira versão.


Em 14/09/2022 09:03, Mateus Felipe C. C. Pinto escreveu:
